nixos/modules/misc/meta.nix

```
@@ -39,6 +39,64 @@ let
     #   { file = "module location"; value = <path/to/doc.xml>; }
     merge = loc: defs: defs;
   };
+
+  /**
+    Custom type for `meta.tests` option.
+  */
+  testsType = lib.mkOptionType {
+    name = "tests";
+    check = value: lib.isFunction value;
+    merge = _loc: defs: rec {
+      byModulePath =
+        lib.zipAttrsWith
+          (
+            _fileName: values:
+            # These are per file, so we don't need to merge definitions.
+            # We could, but it's not implemented yet.
+            assert (lib.length values == 1);
+            lib.head values { inherit (pkgs) nixosTests; }
+          )
+          (
+            map (
+              { file, value, ... }:
+              {
+                "${toString file}" = value;
+              }
+            ) defs
+          );
+      /**
+        Given a list of file names, return a nested attribute set of tests to run, in an efficient lexicographic order.
+      */
+      filter =
+        paths:
+        let
+          pathSet = lib.genAttrs (map toString paths) (_path: null);
+          relevant = builtins.intersectAttrs pathSet byModulePath;
+
+          # Optimize evaluation order by sorting by test name. This way, even
+          # if multiple files reference the same tests, and if you split
+          # evaluation, you can likely resume from a given point in the nested
+          # attrset without encountering already evaluated tests again. (or
+          # split work across multiple processes)
+          /**
+            An attrset structure of <testName>.<fileName> = derivation;
+          */
+          transposed = lib.zipAttrsWith (_k: lib.mergeAttrsList) (
+            lib.mapAttrsToList (fileName: values: lib.mapAttrs (k: v: { "${fileName}" = v; }) values) relevant
+          );
+
+        in
+        transposed;
+
+      /**
+        Given a list of file names, return a list of test derivations for `nix-build`.
+
+        This is like `filter`, but solves the problem that `nix-build` ignores attributes with `.` in their names.
+      */
+      filterForNixBuild = paths: lib.concatMap lib.attrValues (lib.attrValues (filter paths));
```

Would this be more readable as a pipe?

```suggestion
      filterForNixBuild = paths: lib.pipe paths [
        filter
        builtins.attrValues
        (builtins.concatMap builtins.attrValues)
      ];
```

Although, more verbose so maybe not worth it.

---

Alternatively, how about using `collect isDerivation`? Maybe this would be less efficient, but it might convey the intention more clearly:

 ```suggestion
      filterForNixBuild = paths: lib.collect lib.isDerivation (filter paths);
```

---

> but it might convey the intention more clearly:

It somewhat blurs the intent, because it's two layers and not something arbitrarily deep.
I'll add:

```nix
      # Two calls to `attrValues` because we have exactly two layers to flatten.
```